src/megaclient.cpp

```diff
@@ -13705,19 +14081,269 @@ void MegaClient::execmovetosyncdebris()
 
 // we cannot delete the Sync object directly, as it might have pending
 // operations on it
-void MegaClient::delsync(Sync* sync, bool deletecache)
+void MegaClient::delsync(Sync* sync)
 {
     sync->changestate(SYNC_CANCELED);
 
-    sync->setResumable(false);
-
-    if (deletecache && sync->statecachetable)
+    if (sync->statecachetable)
     {
         sync->statecachetable->remove();
         delete sync->statecachetable;
         sync->statecachetable = NULL;
     }
 
+    removeSyncConfig(sync->tag);
+
+    sync->appData = nullptr;
+
+    syncactivity = true;
+}
+
+error MegaClient::removeSyncConfig(int tag)
+{
+    error e = API_OK;
+
+    if (!syncConfigs || !syncConfigs->removeByTag(tag))
+    {
+        LOG_err << "Found no config for tag " << tag << " upon sync removal";
+        e = API_ENOENT;
+    }
+
+    app->sync_removed(tag);
+    return e;
+}
+
+error MegaClient::removeSyncConfigByNodeHandle(mega::handle nodeHandle)
+{
+    auto tag = 0;
+
+    if (!syncConfigs)
+    {
+        LOG_err << "no SyncConfig upon removeSyncConfigByNodeHandle";
+        return API_ENOENT;
+    }
+
+    auto config = syncConfigs?syncConfigs->getByNodeHandle(nodeHandle):nullptr;
+    if (config)
+    {
+        tag = config->getTag();
+        return removeSyncConfig(config->getTag());
+    }
+
+    LOG_err << "Found no config for handle " << nodeHandle << " upon sync removal";
+    return API_ENOENT;
+}
+
+error MegaClient::saveAndUpdateSyncConfig(const SyncConfig *config, syncstate_t newstate, syncerror_t newSyncError)
+{
+    if (syncConfigs)
+    {
+        assert(config);
+        auto newConfig = *config;
+
+        auto isEnabled = [](syncstate_t state, syncerror_t syncError) -> bool
+        {
+            return state != SYNC_CANCELED && (state != SYNC_DISABLED || syncError != NO_SYNC_ERROR );
+        };
+
+        newConfig.setEnabled(isEnabled(newstate, newSyncError));
+        newConfig.setError(newSyncError);
+
+        syncConfigs->insert(newConfig);
+        return API_OK;
+    }
+    return API_ENOENT;
+}
+
+error MegaClient::changeSyncState(const SyncConfig *config, syncstate_t newstate, syncerror_t newSyncError, bool fireDisableEvent)
+{
+    error e = API_OK;
+    assert(config);
+    if (config)
+    {
+        auto isEnabled = [](syncstate_t state, syncerror_t syncError) -> bool
+        {
+            return state != SYNC_CANCELED && (state != SYNC_DISABLED || syncError != NO_SYNC_ERROR );
+        };
+
+        if ( (config->getError() != newSyncError) || (config->getEnabled() != isEnabled(newstate, newSyncError)) ) //has changed
+        {
+            e = saveAndUpdateSyncConfig(config, newstate, newSyncError);
+        }
+    }
+    else
+    {
+        e = API_ENOENT;
+    }
+
```

Comment:
Probably we want to not callback `syncupdate_state()` here `if (e)`. The action has failed already

Reply:
You will still callback if `saveAndUpdateSyncConfig()` fails and returns `API_ENOENT` (you may want to avoid the callback, as in the `else`.
I wonder if `MegaClient::syncConfigs` (should be `mSyncConfigs`, BTW) could be an object, rather than a pointer dynamically allocated, and that way avoid all those `if (!syncConfig)` along the code.

Reply:
ok no callback if `e`.
regarding syncConfigs change. copy constructor is disabled, and that would require changing the initialization, ... too much ado for nothing IMHO ;-)